tools/contract-analyzer/analyzers/number_function_argument_analyzer.go

```diff
@@ -0,0 +1,218 @@
+
+		// If the integer literal is converted to an integer type,
+		// suggest replacing it with a fixed-point literal
+
+		var hintExpression ast.Expression = argument
+
+		// If the target type is not `Int`,
+		// then a static cast is required,
+		// as all integer literals (positive and negative)
+		// are inferred to be of type `Int`
+
+		if !sema.IsSameTypeKind(targetType, sema.IntType) {
+			hintExpression = ast.NewCastingExpression(
+				nil,
+				hintExpression,
+				ast.OperationCast,
+				ast.NewTypeAnnotation(
+					nil,
+					false,
+					ast.NewNominalType(
+						nil,
+						ast.NewIdentifier(
+							nil,
+							targetType.String(),
+							ast.EmptyPosition,
+						),
+						nil,
+					),
+					ast.EmptyPosition,
+				),
+				nil,
+			)
+		}
+
+		return ReplacementHint(hintExpression, location, invocationRange)
+	}
+
+	return nil
+}
+
+func suggestFixedPointLiteralConversionReplacement(
+	argument *ast.FixedPointExpression,
+	location common.Location,
+	targetType sema.Type,
+	invocationRange ast.Range,
+) *analysis.Diagnostic {
+	// If the fixed-point literal is converted to a fixed-point type,
+	// suggest replacing it with a fixed-point literal
+
+	if !sema.IsSameTypeKind(targetType, sema.FixedPointType) {
+		return nil
+	}
+
+	negative := argument.Negative
+	signed := sema.IsSameTypeKind(targetType, sema.SignedFixedPointType)
+
+	if (!negative && !signed) || (negative && signed) {
+		return ReplacementHint(argument, location, invocationRange)
+	}
+
+	return nil
+}
+
+var NumberFunctionArgumentAnalyzer = (func() *analysis.Analyzer {
+
+	elementFilter := []ast.Element{
+		(*ast.IntegerExpression)(nil),
+		(*ast.FixedPointExpression)(nil),
+	}
+
+	return &analysis.Analyzer{
+		Description: "Detects redundant uses of number conversion functions.",
+		Requires: []*analysis.Analyzer{
+			analysis.InspectorAnalyzer,
+		},
+		Run: func(pass *analysis.Pass) interface{} {
+			inspector := pass.ResultOf[analysis.InspectorAnalyzer].(*ast.Inspector)
+
+			location := pass.Program.Location
+			elaboration := pass.Program.Elaboration
+			report := pass.Report
+
+			inspector.Preorder(
+				elementFilter,
+				func(element ast.Element) {
+
+					integerExpression, intOk := element.(*ast.IntegerExpression)
+					fixedPointExpression, fixedOk := element.(*ast.FixedPointExpression)
+					if !intOk && !fixedOk {
+						return
+					}
+
+					var diagnostic *analysis.Diagnostic
+
+					if intOk {
+						argumentData := elaboration.NumberConversionArgumentTypes[integerExpression]
+						diagnostic = suggestIntegerLiteralConversionReplacement(integerExpression, location, argumentData.Type, argumentData.Range)
+					}
+
+					if fixedOk {
+						argumentData := elaboration.NumberConversionArgumentTypes[fixedPointExpression]
+						diagnostic = suggestFixedPointLiteralConversionReplacement(fixedPointExpression, location, argumentData.Type, argumentData.Range)
+					}
+
+					if diagnostic != nil {
+						report(*diagnostic)
+					}
```

Comment:
Maybe use a type switch case statement here instead, only one of the two if-statement branches will ever be executed